Google.Api.Gax.Grpc/Rest/RestMethod.cs

```diff
@@ -82,19 +78,122 @@ internal static RestMethod Create(MethodDescriptor method, JsonParser parser)
         /// <returns>A request with the URI, method and content populated.</returns>
         internal HttpRequestMessage CreateRequest(IMessage protoRequest, string host)
         {
-            string path = _pathPattern.Format(protoRequest);
-            var uri = host is null ? new Uri(path, UriKind.Relative) : new UriBuilder { Host = host, Path = path }.Uri;
-            string jsonContent = _contentFactory?.Invoke(protoRequest);
+            var transcodingResult = Transcode(protoRequest);
+
+            var uriPathWithParams = QueryHelpers.AddQueryString(transcodingResult.UriPath, transcodingResult.QueryStringParameters);
+            var uri = host is null ? new Uri(uriPathWithParams, UriKind.Relative) : new UriBuilder { Host = host, Path = transcodingResult.UriPath }.Uri;
+            
             return new HttpRequestMessage
             {
                 RequestUri = uri,
                 Method = _httpMethod,
-                Content = jsonContent is null ? null : new StringContent(jsonContent, Encoding.UTF8, s_applicationJsonMediaType)
+                Content = transcodingResult.Body is null ? null : new StringContent(transcodingResult.Body, Encoding.UTF8, s_applicationJsonMediaType),
             };
         }
 
-        // TODO: Handle cancellation?
+        /// <summary>
+        /// GRPC Transcoding returns the parameters of the http request, namely
+        ///  - the path of the uri
+        ///  - the query string parameters
+        ///  - the body
+        /// based on the
+        ///  - method descriptor, specifically the list of input message fields, and the `google.api.http` annotation
+        ///  - the request message
+        /// It is described in the AIP-127 (https://google.aip.dev/127) and the proto comments for google.api.HttpRule
+        /// (https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L44-L312)
+        /// </summary>
+        /// <param name="protoRequest">the request message</param>
+        /// <returns>GRPC Transcoding output</returns>
+        private TranscodingOutput Transcode(IMessage protoRequest)
+        {
+            string path = _pathPattern.Format(protoRequest);
+            (bool allFieldsInBody, string bodyName, string bodyResult) = _httpBody switch
+            {
+                "*" => (true, null, protoRequest.ToString()),
+                "" => (false, null, null),
+                string name when _protoMethod.InputType.FindFieldByName(name) is FieldDescriptor field 
+                    => (false, name, field.Accessor.GetValue(protoRequest).ToString()),
+                _ => throw new ArgumentException($"Method {_protoMethod.Name} in service {_protoMethod.Service.Name} has a body parameter {_httpBody} in the 'google.api.http. annotation which is not a field in {_protoMethod.InputType.Name}")
```

Comment:
Is the period after http meant to be an apostrophe?

Reply:
👍